source/common/websocket/codec.h

```diff
@@ -0,0 +1,146 @@
+  // decoding succeeded (returns true). If the input is not sufficient to make a
+  // complete WebSocket frame, it will be buffered in the decoder. If a decoding
+  // error happened, the input buffer remains unchanged.
+  // @param input supplies the binary octets wrapped in a WebSocket frame.
+  // @param output supplies the buffer to store the decoded data.
+  // @return bool whether the decoding succeeded or not.
+  absl::optional<std::vector<Frame>> decode(Buffer::Instance& input);
+
+protected:
+  // Decodes the given buffer with WebSocket data frames and updates the frame count.
+  // Invokes visitor callbacks for each frame in the following sequence:
+  // ----------------------------------------------------------------------------------
+  //  frameStart frameMaskFlag frameMaskingKey? frameDataStart frameData* frameDataEnd
+  // ----------------------------------------------------------------------------------
+  // If frameStart returns false, then the decoder aborts.
+  // Returns the increase in the frame count.
+  uint64_t inspect(const Buffer::Instance& input);
+
+  bool frameStart(uint8_t);
+  void frameMaskFlag(uint8_t);
+  void frameMaskingKey();
+  void frameDataStart();
+  void frameData(const uint8_t*, uint64_t);
+  void frameDataEnd();
+
+  void doDecodeMaskFlagAndLength(const uint8_t, const uint8_t*& mem, uint64_t&);
+  void doDecodeExtendedLength(const uint8_t c, const uint8_t*&, uint64_t&);
+  void doDecodeMaskingKey(const uint8_t c, const uint8_t*&, uint64_t&);
+  void doDecodePayload(const Buffer::RawSlice&, const uint8_t*&, uint64_t&);
+
+private:
+  // Current state of the frame that is being processed.
+  enum class State {
+    // Decoding the first byte. Waiting for decoding the final frame flag (1 bit)
+    // and reserved flags (3 bits) and opcode (4 bits) of the WebSocket data frame.
+    KFrameHeaderFlagsAndOpcode,
+    // Decoding the second byte. Waiting for decoding the mask flag (1 bit) and
+    // length/length flag (7 bit) of the WebSocket data frame.
+    KFrameHeaderMaskFlagAndLength,
+    // Waiting for decoding the extended length of the frame if length read previously
+    // is either 126 or 127. Respectively 2 bytes or 8 bytes will be decoded from the
+    // WebSocket data frame.
+    KFrameHeaderExtendedLength,
+    // Waiting for decoding the masking key (4 bytes) only if the mask bit is set.
+    KFrameHeaderMaskingKey,
+    // Waiting for decoding the payload (both extension data and application data).
+    KFramePayload,
+  };
+  // Current frame that is being decoded.
+  Frame frame_;
+  // Data holder for successfully decoded frames.
+  absl::optional<std::vector<Frame>> output_ = absl::nullopt;
+  bool decoding_error_ = false;
+  State state_ = State::KFrameHeaderFlagsAndOpcode;
+  uint64_t total_frames_count_ = 0;
+  uint8_t length_of_extended_length_ = 0;
+  uint32_t masking_key_ = 0;
+  uint64_t length_ = 0;
```

Comment:
Is this field different from `frame_.payload_length_`? If not, let's remove it.

Reply:
`length_` is the processing length, `length_` will get the final value after enough bytes are consumed and at the end of the payload is also processed, then we set the `frame_.payload_length_` to its value.